repos/system_upgrade/el7toel8/actors/selinux/selinuxcontentscanner/actor.py

```diff
@@ -0,0 +1,223 @@
+
+
+class SELinuxContentScanner(Actor):
+    '''
+    Scan the system for any SELinux customizations
+
+    Find SELinux policy customizations (custom policy modules and changes
+    introduced by semanage) and save them in SELinuxModules and SELinuxCustom
+    models. Customizations that are incompatible with SELinux policy on RHEL-8
+    are removed.
+    '''
+    name = 'selinuxcontentscanner'
+    consumes = (SELinuxFacts, )
+    produces = (SELinuxModules, SELinuxCustom, SELinuxRequestRPMs, RpmTransactionTasks,)
+    tags = (FactsPhaseTag, IPUWorkflowTag, )
+
+    def process(self):
+        # exit if SELinux is disabled
+        for fact in self.consume(SELinuxFacts):
+            if not fact.enabled:
+                return
+
+        (semodule_list, request_rpms) = self.getSELinuxModules()
+
+        self.produce(SELinuxModules(modules=semodule_list))
+        self.produce(
+            RpmTransactionTasks(
+                to_install = request_rpms[1],
+                # possibly not necessary - dnf should not remove RPMs (that exist in both RHEL 7 and 8) durign update
+                to_keep = request_rpms[0]
+            )
+        )
+        # this is produced so that we can later verify that the RPMs are present after upgrade
+        self.produce(
+            SELinuxRequestRPMs(
+                to_install = request_rpms[1],
+                to_keep = request_rpms[0]
+            )
+        )
+
+        semanage_removed = []
+        semanage_valid = []
+        try:
+            # Collect SELinux customizations and select the ones that
+            # can be reapplied after the upgrade
+            semanage = run(['semanage', 'export'], split=True)
+            for line in semanage.get("stdout", ""):
+                for setype in REMOVED_TYPES_:
+                    if setype in line:
+                        semanage_removed.append(line)
+                        break
+                else:
+                    semanage_valid.append(line)
+
+        except CalledProcessError:
+            return
+
+        self.produce(
+            SELinuxCustom(
+                commands=semanage_valid,
+                removed=semanage_removed
+            )
+        )
+
+    def checkModule(self, name):
+        '''
+        Check if given module contains one of removed types.
+
+        If so, comment out corresponding lines and return them.
+        The function expects a text file "$name" containing cil policy
+        to be present in the current directory.
+        '''
+        try:
+            removed = run(['grep', '-w', '-E', REMOVED_TYPES, name], split=True)
+            run(['sed', '-i', SED_COMMAND, name])
+            return removed.get("stdout", "")
```

Comment:
Inconsistent return, [] in case of Exception and ""

Reply:
Not true. 
run returns:
'stdout': result['stdout'].splitlines()
which is a list